libimage/manifests/manifests.go

```diff
@@ -455,6 +634,222 @@ func (l *list) Add(ctx context.Context, sys *types.SystemContext, ref types.Imag
+			}
+			defer f.Close()
+
+			// Hang on to a copy of the first 512 bytes, but digest the whole thing.
+			digester := digest.Canonical.Digester()
+			writeCounter := ioutils.NewWriteCounter(digester.Hash())
+			var detectableData bytes.Buffer
+			_, err = io.CopyN(writeCounter, io.TeeReader(f, &detectableData), 512)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return fmt.Errorf("reading %q to determine its digest: %w", file, err)
+			}
+			if err == nil {
+				if _, err := io.Copy(writeCounter, f); err != nil {
+					return fmt.Errorf("reading %q to determine its digest: %w", file, err)
+				}
+			}
+			fileDigests[absFile] = digester.Digest()
+
+			// If one wasn't specified, figure out what the MediaType should be.
+			title := filepath.Base(absFile)
+			layerMediaType := options.LayerMediaType
+			if layerMediaType == nil {
+				if index := strings.LastIndex(title, "."); index != -1 {
+					// File's basename has an extension, try to use a shortcut.
+					tmp := mime.TypeByExtension(title[index:])
+					if tmp != "" {
+						layerMediaType = &tmp
+					}
+				}
+				if layerMediaType == nil {
+					// File's basename has no extension or didn't map to a type, look at the contents we saved.
+					tmp := http.DetectContentType(detectableData.Bytes())
+					layerMediaType = &tmp
+				}
+				if layerMediaType != nil {
+					// Strip off any parameters, since we only want the type name.
+					if parsedMediaType, _, err := mime.ParseMediaType(*layerMediaType); err == nil {
+						layerMediaType = &parsedMediaType
+					}
+				}
+			}
+
+			// Only set an ArtifactType for the layer if one was specified.
+			layerArtifactType := options.LayerArtifactType
+			if layerArtifactType == nil {
+				tmp := ""
+				layerArtifactType = &tmp
+			}
+			descriptor := v1.Descriptor{
+				MediaType:    *layerMediaType,
+				ArtifactType: *layerArtifactType,
+				Digest:       fileDigests[absFile],
+				Size:         writeCounter.Count,
+			}
+			// Add a title annotation unless we were explicitly told not to.
+			if !options.ExcludeTitles {
+				descriptor.Annotations = map[string]string{
+					v1.AnnotationTitle: title,
+				}
+			}
+			layers = append(layers, descriptor)
+			return nil
+		}(); err != nil {
+			return "", err
+		}
+	}
+
+	// Unless we were explicitly told otherwise, default to the normal image MediaType.
+	mediaType := v1.MediaTypeImageManifest
+	if options.ManifestMediaType != nil && *options.ManifestMediaType != "" {
+		mediaType = *options.ManifestMediaType
+	}
+
+	// Unless we were told what this is, use the default that ORAS uses.
+	artifactType := "application/vnd.unknown.artifact.v1"
+	if options.ManifestArtifactType != nil {
+		artifactType = *options.ManifestArtifactType
+	}
+
+	// Unless we were explicitly told otherwise, default to an empty config blob.
+	configDescriptor := &v1.DescriptorEmptyJSON
+	if options.ConfigDescriptor != nil {
+		configDescriptor = options.ConfigDescriptor
```

Comment:
Is this sufficient to actually make the _data_ of the config available?

Layers are accessible via `l.artifacts.Files` , where is the config coming from?

Reply:
Hmm, no, it isn't.  I had expected that it would still be an inlined `Data` field, but you're right that it's not really enough to expect that.

Reply:
Just a descriptor with `Data` might work … except that _right now_ c/image doesn’t know about `Data`, and relies on `ImageSource.GetBlob(digest)` only.

Fixing that in `c/image/copy` should be fairly easy, and then this can work. OTOH if we force callers to set `.Data`, that `.Data` will (probably?) always end up in the generated manifest written to the destination, IOW it could be inconvenient for very large config blobs.

Reply:
I figured inlined `Data` was already on your checklist.  It's easy enough at this stage to add a reader or filename field (or both) to the options struct for cases where the config descriptor is set, its size is not zero, and it includes no data.

Reply:
For the record, it’s one of the checklist items in https://github.com/containers/image/issues/2030 . At this point not something I’m being asked to work on with urgency.

Reply:
Yup, that's the one.  Since we're on the producing end, we can default to generating the blobs for zero-length and "{}" data on the fly.  I don't know how that affects the prioritization, then.